Sources/Citadel/SFTP/SFTPBasicEnums.swift

```diff
@@ -0,0 +1,153 @@
+    public init?(rawValue: UInt32) {
+        switch rawValue {
+        case 3: self = .v3
+        case let n: self = .unsupported(n)
+        }
+    }
+}
+
+public enum SFTPMessageType: UInt8 {
+    case initialize = 1
+    case version = 2
+    case openFile = 3
+    case closeFile = 4
+    case read = 5
+    case write = 6
+    case lstat = 7
+    case fstat = 8
+    case setstat = 9
+    case fsetstat = 10
+    case opendir = 11
+    case readdir = 12
+    case remove = 13
+    case mkdir = 14
+    case rmdir = 15
+    case realpath = 16
+    case stat = 17
+    case readlink = 18
+    case symlink = 20
+    
+    case status = 101
+    case handle = 102
+    case data = 103
+    case name = 104
+    case attributes = 105
+    
+    case extended = 200
+    case extendedReply = 201
+    
+    public var description: String {
+        switch self {
+        case .initialize: return "SSH_FXP_INIT"
+        case .version: return "SSH_FXP_VERSION"
+        case .openFile: return "SSH_FXP_OPEN"
+        case .closeFile: return "SSH_FXP_CLOSE"
+        case .read: return "SSH_FXP_READ"
+        case .write: return "SSH_FXP_WRITE"
+        case .lstat: return "SSH_FXP_LSTAT"
+        case .fstat: return "SSH_FXP_FSTAT"
+        case .setstat: return "SSH_FXP_SETSTAT"
+        case .fsetstat: return "SSH_FXP_FSETSTAT"
+        case .opendir: return "SSH_FXP_OPENDIR"
+        case .readdir: return "SSH_FXP_READDIR"
+        case .remove: return "SSH_FXP_REMOVE"
+        case .mkdir: return "SSH_FXP_MKDIR"
+        case .rmdir: return "SSH_FXP_RMDIR"
+        case .realpath: return "SSH_FXP_REALPATH"
+        case .stat: return "SSH_FXP_STAT"
+        case .readlink: return "SSH_FXP_READLINK"
+        case .symlink: return "SSH_FXP_SYMLINK"
+
+        case .status: return "SSH_FXP_STATUS"
+        case .handle: return "SSH_FXP_HANDLE"
+        case .data: return "SSH_FXP_DATA"
+        case .name: return "SSH_FXP_NAME"
+        case .attributes: return "SSH_FXP_ATTRS"
+        
+        case .extended: return "SSH_FXP_EXTENDED"
+        case .extendedReply: return "SSH_FXP_EXTENDED_REPLY"
+        }
+    }
+}
+
+public enum SFTPStatusCode: RawRepresentable, Hashable, CustomDebugStringConvertible {
+    case ok
+    case eof
+    case noSuchFile
+    case permissionDenied
+    case failure
+    case badMessage
+    case noConnection
+    case connectionLost
+    case unsupportedOperation
+    case unknown(UInt32)
+    
+    public var rawValue: UInt32 {
+        switch self {
+        case .ok: return 0
+        case .eof: return 1
+        case .noSuchFile: return 2
+        case .permissionDenied: return 3
+        case .failure: return 4
+        case .badMessage: return 5
+        case .noConnection: return 6
+        case .connectionLost: return 7
+        case .unsupportedOperation: return 8
+        case .unknown(let value): return value
+        }
+    }
+    
+    public init?(rawValue: UInt32) {
```

Comment:
Can we make this non-failable? And get rid of some force unwrap operators that work around this non-failable failable initialiser?

Reply:
Not without losing `RawRepresentable` conformance; the compiler won't auto-synthesize thanks to the associated value on the `.unsupported` case.

From what I've seen, most libraries deal with the "we want to enumerate the well-known values but still be able to represent unknown ones" problem by using a `struct` instead of an `enum` (like the way an `OptionSet` works  but without the bitwise stuff). But that solution sacrifices all the benefits of an algebraic sum type (an enum), and since the compiler won't let me say the type is a `UInt32`-backed `enum` directly, manually implementing `RawRepresentable` allows the full range of possible values while keeping things like the compiler checks for exhaustive `switch` cases, the clear indication of which values are "valid", and so on. Compare to a `struct` version:

```swift
public struct SFTPProtocolVersion: RawRepresentable, Hashable, Comparable {
    public var rawValue: UInt32
    public init?(rawValue: UInt32) { self.rawValue = rawValue }
    public init(rawValue: UInt32) { self.rawValue = rawValue }
    
    public static let v3: Self { .init(rawValue: 3)! }
    public static func unsupported(_ rawValue: RawValue) -> Self {  .init(rawValue: rawValue)! }
}
```
Yeah, it's shorter than the enum, but most of that is comments and formatting anyway! 😅 Though I should add the non-optional copy of the raw value initializer to the `enum` for convenience's sake.

(Okay, I admit it - my biggest problem with using a `struct` for this kind of case is pretty much that it's conceptually wrong; enums are specifically intended to represent a constrained set of valid states, structs are just bags of other types >.<. It doesn't _really_ matter all that much, especially since you need the `RawRepresentable` conformance either way to keep it from being a pain in the rear to use.)

I consider the lack of a `LosslessRawRepresentable` protocol (same as `RawRepresentable` but without the failable init) to be a design flaw we got stuck with because Swift didn't want to admit to itself that it forces authors to implement awkward wrapper types all the time (and has nothing like `newtype` to enforce type safety for simple scalar values); without that use case, a `LosslessRawRepresentable` value would just be a less efficient representation of an existing type. Except that use case exists a lot (especially thanks to limitations of generics) and network protocols especially tend to have a need to be able to represent "not well-known but not programmer error either" value types...

Meandering ramble showing off my lack of knowledge of category theory notwithstanding 😅, I'll add the non-failable overload and keep the existing one for protocol conformance, and see how many of the `!`s I can get rid of.